packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts

```diff
@@ -475,3 +477,54 @@ export const mergeColumnsState =
     ...state,
     columns: columnsState,
   });
+
+export function getFirstNonSpannedColumnToRender({
+  firstColumnToRender,
+  apiRef,
+  firstRowToRender,
+  lastRowToRender,
+}: {
+  firstColumnToRender: number;
+  apiRef: React.MutableRefObject<GridApiCommon>;
+  firstRowToRender: number;
+  lastRowToRender: number;
+}) {
+  let firstNonSpannedColumnToRender = firstColumnToRender;
+  const visibleSortedRows = gridVisibleSortedRowEntriesSelector(apiRef);
+  const renderedRows = visibleSortedRows.slice(firstRowToRender, lastRowToRender);
```

Comment:
`Array.slice` has complexity O(n). Inside `getRows` you already have a `renderedRows` which you can pass to here.

Reply:
> Inside getRows you already have a renderedRows

There are more places that use `getFirstNonSpannedColumnToRender` indirectly (through `getFirstColumnIndexToRender`)

I can use `for` loop based on `firstRowToRender` and `lastRowToRender` instead